source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h

```diff
@@ -116,6 +138,267 @@ class UdpLoadBalancerContext : public Upstream::LoadBalancerContextBase {
+
+  /**
+   * Called when an event was raised by the upstream.
+   * @param event the event.
+   */
+  virtual void onUpstreamEvent(Network::ConnectionEvent event) PURE;
+
+  /**
+   * Called when the upstream buffer went above high watermark.
+   */
+  virtual void onAboveWriteBufferHighWatermark() PURE;
+
+  /**
+   * Called when the upstream buffer went below high watermark.
+   */
+  virtual void onBelowWriteBufferLowWatermark() PURE;
+
+  /**
+   * Called when an event was raised by the upstream.
+   * @param data the data received from the upstream.
+   * @param end_stream indicates if the received data is ending the stream.
+   */
+  virtual void onUpstreamData(Buffer::Instance& data, bool end_stream) PURE;
+};
+
+class HttpUpstreamImpl : public HttpUpstream, protected Http::StreamCallbacks {
+public:
+  HttpUpstreamImpl(UpstreamTunnelCallbacks& upstream_callbacks, const UdpTunnelingConfig& config,
+                   StreamInfo::StreamInfo& downstream_info)
+      : response_decoder_(*this), upstream_callbacks_(upstream_callbacks),
+        downstream_info_(downstream_info), tunnel_config_(config) {}
+  ~HttpUpstreamImpl() override;
+
+  Http::ResponseDecoder& responseDecoder() { return response_decoder_; }
+  void setRequestEncoder(Http::RequestEncoder& request_encoder, bool is_ssl);
+  void setTunnelCreationCallbacks(TunnelCreationCallbacks& callbacks) {
+    tunnel_creation_callbacks_.emplace(callbacks);
+  }
+
+  // HttpUpstream
+  void encodeData(Buffer::Instance& data) override;
+  void onDownstreamEvent(Network::ConnectionEvent event) override {
+    if (event == Network::ConnectionEvent::LocalClose ||
+        event == Network::ConnectionEvent::RemoteClose) {
+      resetEncoder(event, /*by_downstream=*/true);
+    }
+  };
+
+  // Http::StreamCallbacks
+  void onResetStream(Http::StreamResetReason, absl::string_view) override {
+    resetEncoder(Network::ConnectionEvent::LocalClose);
+  }
+
+  void onAboveWriteBufferHighWatermark() override {
+    upstream_callbacks_.onAboveWriteBufferHighWatermark();
+  }
+
+  void onBelowWriteBufferLowWatermark() override {
+    upstream_callbacks_.onBelowWriteBufferLowWatermark();
+  }
+
+private:
+  class ResponseDecoder : public Http::ResponseDecoder {
+  public:
+    ResponseDecoder(HttpUpstreamImpl& parent) : parent_(parent) {}
+
+    // Http::ResponseDecoder
+    void decodeMetadata(Http::MetadataMapPtr&&) override {}
+    void decode1xxHeaders(Http::ResponseHeaderMapPtr&&) override {}
+    void dumpState(std::ostream& os, int indent_level) const override {
+      DUMP_STATE_UNIMPLEMENTED(ResponseDecoder);
+    }
+
+    void decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override {
+      bool is_valid_response = Http::CodeUtility::is2xx(Http::Utility::getResponseStatus(*headers));
+      bool is_capsule_protocol = Http::HeaderUtility::isCapsuleProtocol(*headers);
+
+      if (!is_valid_response || !is_capsule_protocol || end_stream) {
```

Comment:
The CapsuleProtocol header isn't required, no need to check for it here

Reply:
done